main/solution/machine-images/config/infra/files/rstudio/check-idle

```diff
@@ -1,42 +1,53 @@
 #!/usr/bin/env python
 
-
 import os
-import time
+import datetime as dt
+import dateutil.parser as parser
 
 
 LAST_ACTIVE_TIME_FILENAME = '/tmp/last-active-rsession-time'
+ACCESS_LOG_FILENAME = '/var/log/nginx/access.log'
 
-MAX_IDLE_SECONDS = 3600
-
+# To disable auto-stop for RStudio, enter 0 as the value below for MAX_IDLE_MINUTES
+MAX_IDLE_MINUTES = 60
 
-def get_uptime():
-    return float(open('/proc/uptime').read().split()[0])
 
+def get_activity_count():
+    non_ping_count = 0
+    with open(ACCESS_LOG_FILENAME) as access_log:
+        for line in access_log:
+            # nginx sometimes pings the RStudio instance
+            # which does not indicate user activity
+            if "POST /events/get_events" not in line:
+                non_ping_count += 1
+    return (non_ping_count)
 
-def is_rsession(pid):
-    try:
-        cmdline = open(os.path.join('/proc', pid, 'cmdline'), 'rb').read()
-        return 'rsession' in cmdline
-    except IOError:
-        return False
 
+current_time = dt.datetime.now()
+latest_activity_count = get_activity_count()
 
-all_pids = [p for p in os.listdir('/proc') if p.isdigit()]
-rsession_pids = [p for p in all_pids if is_rsession(p)]
+if not os.path.exists(LAST_ACTIVE_TIME_FILENAME):
+    open(LAST_ACTIVE_TIME_FILENAME, 'w').write(
+        str(latest_activity_count) + '\n' + str(current_time))
 
-num_sessions = len(rsession_pids)
+# These will be overwritten by the stored file values
+last_activity_count = 0
+last_active_time = dt.datetime.now()
 
-current_time = time.time()
+with open(LAST_ACTIVE_TIME_FILENAME) as fh:
+    last_activity_count = int(fh.readline())
+    for line in fh:
+        pass
+    last_active_time = parser.parse(line)
 
-if num_sessions or not os.path.exists(LAST_ACTIVE_TIME_FILENAME):
-    open(LAST_ACTIVE_TIME_FILENAME, 'w').write(str(current_time))
-    print('{0} RStudio sessions active at {1}'.format(num_sessions, current_time))
+# Compare with old_activity_count
+# If higher than old count, update file with time and count
+if latest_activity_count > last_activity_count:
+    open(LAST_ACTIVE_TIME_FILENAME, 'w').write(
+        str(latest_activity_count) + '\n' + str(current_time))
 else:
-    last_active_time = float(open(LAST_ACTIVE_TIME_FILENAME).read())
-    idle_time = current_time - last_active_time
-    uptime = get_uptime()
-    print('RStudio has been idle for {0} seconds and system uptime is {1}'.format(idle_time, uptime))
-    if idle_time >= MAX_IDLE_SECONDS and uptime >= MAX_IDLE_SECONDS:
+    # If time difference between file time and now exceeds MAX_IDLE_MINUTES, shutdown
+    idle_minutes = (current_time - last_active_time).total_seconds()/60
+    if MAX_IDLE_MINUTES != 0 and idle_minutes >= MAX_IDLE_MINUTES:
```

Comment:
Minor - Is there a way to quit the file execution early on? If MAX_IDLE_MINUTES == 0, the file don't really need to be ran.

Reply:
Good point. Making that change.